src/idom/utils.py

```diff
@@ -56,6 +62,52 @@ def __repr__(self) -> str:
         return f"{type(self).__name__}({current})"
 
 
+def vdom_to_html(value: str | VdomDict) -> str:
+    """Convert a VDOM dictionary into an HTML string
+
+    Only the following keys are translated to HTML:
+
+    - ``tagName``
+    - ``attributes``
+    - ``children`` (must be strings or more VDOM dicts)
+    """
+
+    if isinstance(value, str):
+        return value
+
+    try:
+        tag = value["tagName"]
+    except TypeError as error:  # pragma: no cover
+        raise TypeError(f"Expected a VDOM dictionary or string, not {value}") from error
+
+    attributes = " ".join(
+        _vdom_to_html_attr(k, v) for k, v in value.get("attributes", {}).items()
+    )
+
+    if attributes:
+        assert tag, "Element frament may not contain attributes"
+        attributes = f" {attributes}"
+
+    children = "".join(
+        vdom_to_html(cast("VdomDict | str", c))
+        if isinstance(c, (dict, str))
+        else html_escape(str(c))
+        for c in value.get("children", ())
+    )
+
+    return (
+        (
+            f"<{tag}{attributes}>{children}</{tag}>"
+            if children
+            # To be safe we mark elements without children as self-closing.
+            # https://html.spec.whatwg.org/multipage/syntax.html#foreign-elements
+            else (f"<{tag}{attributes} />" if attributes else f"<{tag}/>")
```

Comment:
This does not properly handle situations like `<img>` tags.

Can potentially considering `lxml` to handle this automatically.

Reply:
So there's some value in `lxml` for this, but the way LXML handles text turns out to make things more complicated on some ways.